### ISSUE TITLE
Upgrade to v3.9 fails with missing image

### ISSUE THREAD

**Issue by clcollins** [openshift/openshift-ansible#8025]
#### Description
During upgrade from v3.7 to v3.9, Ansible run fails to pull etcd image

##### Version

* Your ansible version per `ansible --version`

If you're operating from a **git clone**:

* The output of `git describe`

If you're running from playbooks installed via RPM

* The output of `rpm -q openshift-ansible`

Place the output between the code block below:

```
[chris@krang openshift-ansible] $ ansible --version
ansible 2.5.0
  config file = /home/chris/Projects/duke-openshift/dev_cluster/openshift-ansible/ansible.cfg
  configured module search path = [u'/home/chris/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Mar 14 2018, 13:36:31) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
->[release-3.9 ok]<-
[chris@krang openshift-ansible] $ git describe
openshift-ansible-3.9.22-1-17-g013747080

```

##### Steps To Reproduce
1. ansible-playbook -i ../inventory.yml playbooks/byo/openshift-cluster/upgrades/v3_9/upgrade.yml

##### Expected Results

Expected to complete upgrade to v3.9 via "automated in-place upgrade" described here:
https://docs.openshift.org/3.7/upgrading/automated_upgrades.html#running-upgrade-playbooks

##### Observed Results
Upgrade runs until attempting to pull new etcd image:

```
[chris@krang openshift-ansible] $ ansible-playbook -i ../inventory.yml playbooks/byo/openshift-cluster/upgrades/v3_9/upgrade.yml

<snip>

TASK [etcd : Pull new etcd image] **********************************************
Wednesday 18 April 2018  12:24:08 -0400 (0:00:00.103)       0:05:33.455 ******* 
fatal: [<REDACTED>]: FAILED! => {"changed": true, "cmd": ["docker", "pull", "registry.fedoraproject.org/f26/etcd:3.2.7"], "delta": "0:00:00.894557", "end": "2018-04-18 16:24:09.662132", "msg": "non-zero return code", "rc": 1, "start": "2018-04-18 16:24:08.767575", "stderr": "Error: image f26/etcd:3.2.7 not found", "stderr_lines": ["Error: image f26/etcd:3.2.7 not found"], "stdout": "Trying to pull repository registry.fedoraproject.org/f26/etcd ... \nPulling repository registry.fedoraproject.org/f26/etcd", "stdout_lines": ["Trying to pull repository registry.fedoraproject.org/f26/etcd ... ", "Pulling repository registry.fedoraproject.org/f26/etcd"]}

PLAY RECAP *********************************************************************
localhost                  : ok=11   changed=0    unreachable=0    failed=0   
<REDACTED> : ok=105  changed=12   unreachable=0    failed=1   

Wednesday 18 April 2018  12:24:09 -0400 (0:00:01.393)       0:05:34.849 ******* 
=============================================================================== 
Run health checks (upgrade) ------------------------------------------- 256.12s
Verify containers are available for upgrade ----------------------------- 6.77s
Ensure Master is running ------------------------------------------------ 5.98s
etcd : Check current etcd disk usage ------------------------------------ 5.50s
Check if iptables is running -------------------------------------------- 5.00s
Gather Cluster facts ---------------------------------------------------- 2.96s
openshift_facts --------------------------------------------------------- 2.82s
Initialize openshift.node.sdn_mtu --------------------------------------- 2.73s
Gathering Facts --------------------------------------------------------- 2.41s
openshift_version : Set precise containerized version to configure if openshift_release specified --- 2.20s
openshift_facts --------------------------------------------------------- 1.96s
etcd : Configure etcd profile.d aliases --------------------------------- 1.40s
etcd : Pull new etcd image ---------------------------------------------- 1.39s
etcd : Record containerized etcd version (docker) ----------------------- 1.29s
openshift_etcd_facts : Check for CA indicator files --------------------- 1.26s
etcd : Generate etcd backup --------------------------------------------- 1.09s
Read master storage backend setting ------------------------------------- 0.97s
container_runtime : Check if Docker is installed ------------------------ 0.80s
Determine Atomic Host Docker Version ------------------------------------ 0.78s
etcd : Detecting Atomic Host Operating System --------------------------- 0.77s


Failure summary:


  1. Hosts:    <REDACTED>
     Play:     Upgrade containerized hosts to 3.2.7
     Task:     Pull new etcd image
     Message:  non-zero return code

```

##### Additional Information

Provide any additional information which may help us diagnose the
issue.

* Operating System:  Red Hat Enterprise Linux Atomic Host release 7.5
* I don't believe the inventory file contains any special non-standard setup, other than:
* * Atomic Hosts
* * `containerized=true` in inventory file

##### Question
Should the OpenShift version information be changed in the inventory file PRIOR to upgrading? eg:
```
openshift_release=v3.7
openshift_image_tag=v3.7.1
openshift_version=3.7.1
```
Thanks for any help.

**Comment by MansM** [comment]
do you use oreg_url or additional registries?

add this to your inventory:
etcd_image: yourregistry/rhel7/etcd
openshift_cli_image: yourregistry/openshift/origin
osm_image: yourregistry/openshift/origin
osn_image: yourregistry/openshift/node
osn_ovs_image: yourregistry/openshift/openvswitch
openshift_web_console_image_name: yourregistry/openshift/origin-web-console

**Comment by clcollins** [comment]
I did have an additional registry configured, but not one that contained any of those images - it's for internal images only.

I commented that one out, and re-ran, and had the same issue.

I'm also not sure if it's right or not, but I tried setting the images you suggested, minus the "yourregistry/" prefix, and received:

```
FAILED! => {"changed": true, "checks": {"disk_availability": {"skipped": true, "skipped_reason": "Disabled by user request"}, "docker_image_availability": {"changed": true, "failed": true, "failures": [["OpenShiftCheckException", "One or more required container images are not available:\n    rhel7/etcd\nChecked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>\nDefault registries searched: docker.io\n"]], "msg": "One or more required container images are not available:\n    rhel7/etcd\nChecked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>\nDefault registries searched: docker.io\n"}, "memory_availability": {}}, "msg": "One or more checks failed", "playbook_context": "upgrade"}
```

So it looks as though it only looked on docker.io, and not registry.access.redhat.com.  I verified I could manually pull the image from registry.access.redhat.com, and then added it as the prefix in my inventory file, but that fails too:

```
FAILED! => {"changed": true, "checks": {"disk_availability": {"skipped": true, "skipped_reason": "Disabled by user request"}, "docker_image_availability": {"changed": true, "failed": true, "failures": [["OpenShiftCheckException", "One or more required container images are not available:\n    registry.access.redhat.com/openshift/node,\n    registry.access.redhat.com/openshift/openvswitch,\n    registry.access.redhat.com/openshift/origin\nChecked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>\nDefault registries searched: docker.io\n"]], "msg": "One or more required container images are not available:\n    registry.access.redhat.com/openshift/node,\n    registry.access.redhat.com/openshift/openvswitch,\n    registry.access.redhat.com/openshift/origin\nChecked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>\nDefault registries searched: docker.io\n"}, "memory_availability": {}}, "msg": "One or more checks failed", "playbook_context": "upgrade"}
```

It doesn't look like it's parsing the RH Registry as a registry - it's still looking at Docker.io.

**Comment by clcollins** [comment]
Just relized that the rhel7/etcd image is the only one that comes from registry.access.redhat.com, so I updated the inventory like so:

```
etcd_image=registry.access.redhat.com/rhel7/etcd
openshift_cli_image=openshift/origin
osm_image=openshift/origin
osn_image=openshift/node
osn_ovs_image=openshift/openvswitch
openshift_web_console_image_name=openshift/origin-web-console
```

It now fails with new message, that look like they're pulling images with the v3.7.1 tags:

```
fatal: [<REDACTED>]: FAILED! => {"changed": true, "checks": {"disk_availability": {"skipped": true, "skipped_reason": "Disabled by user request"}, "docker_image_availability": {"changed": true, "failed": true, "failures": [["OpenShiftCheckException", "One or more required container images are not available:\n    cockpit/kubernetes:latest,\n    openshift/node,\n    openshift/openvswitch,\n    openshift/origin,\n    openshift/origin-deployer:v3.7.1,\n    openshift/origin-docker-registry:v3.7.1,\n    openshift/origin-haproxy-router:v3.7.1\nChecked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>\nDefault registries searched: docker.io\nFailed connecting to: docker.io\n"]], "msg": "One or more required container images are not available:\n    cockpit/kubernetes:latest,\n    openshift/node,\n    openshift/openvswitch,\n    openshift/origin,\n    openshift/origin-deployer:v3.7.1,\n    openshift/origin-docker-registry:v3.7.1,\n    openshift/origin-haproxy-router:v3.7.1\nChecked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>\nDefault registries searched: docker.io\nFailed connecting to: docker.io\n"}, "memory_availability": {}}, "msg": "One or more checks failed", "playbook_context": "upgrade"}

NO MORE HOSTS LEFT ************************************************************************************************************************

PLAY RECAP ********************************************************************************************************************************
localhost                  : ok=11   changed=0    unreachable=0    failed=0   
<REDACTED> : ok=31   changed=1    unreachable=0    failed=1   

Thursday 19 April 2018  14:24:01 -0400 (0:01:47.442)       0:02:28.779 ******** 
=============================================================================== 
Run health checks (upgrade) ------------------------------------------------------------------------------------------------------ 107.44s
Gather Cluster facts -------------------------------------------------------------------------------------------------------------- 10.96s
Gathering Facts -------------------------------------------------------------------------------------------------------------------- 5.60s
Initialize openshift.node.sdn_mtu -------------------------------------------------------------------------------------------------- 5.07s
slurp ------------------------------------------------------------------------------------------------------------------------------ 1.89s
openshift_repos : openshift_repos detect ostree ------------------------------------------------------------------------------------ 1.88s
Detecting Operating System from ostree_booted -------------------------------------------------------------------------------------- 1.75s
Determine Atomic Host Docker Version ----------------------------------------------------------------------------------------------- 1.72s
stat ------------------------------------------------------------------------------------------------------------------------------- 1.70s
Check if iptables is running ------------------------------------------------------------------------------------------------------- 1.09s
Evaluate oo_nodes_to_upgrade ------------------------------------------------------------------------------------------------------- 0.55s
openshift_sanitize_inventory : Check for usage of deprecated variables ------------------------------------------------------------- 0.31s
set openshift_deployment_type if unset --------------------------------------------------------------------------------------------- 0.28s
Evaluate groups - g_master_hosts or g_new_master_hosts required -------------------------------------------------------------------- 0.26s
Evaluate oo_all_hosts -------------------------------------------------------------------------------------------------------------- 0.25s
openshift_sanitize_inventory : conditional_set_fact -------------------------------------------------------------------------------- 0.25s
openshift_sanitize_inventory : conditional_set_fact -------------------------------------------------------------------------------- 0.25s
set_fact --------------------------------------------------------------------------------------------------------------------------- 0.25s
initialize_facts set fact openshift_is_atomic and openshift_is_containerized ------------------------------------------------------- 0.25s
openshift_sanitize_inventory : Assign deprecated variables to correct counterparts ------------------------------------------------- 0.24s


Failure summary:


  1. Hosts:    <REDACTED>
     Play:     OpenShift Health Checks
     Task:     Run health checks (upgrade)
     Message:  One or more checks failed
     Details:  check "docker_image_availability":
               One or more required container images are not available:
                   cockpit/kubernetes:latest,
                   openshift/node,
                   openshift/openvswitch,
                   openshift/origin,
                   openshift/origin-deployer:v3.7.1,
                   openshift/origin-docker-registry:v3.7.1,
                   openshift/origin-haproxy-router:v3.7.1
               Checked with: skopeo inspect [--tls-verify=false] [--creds=<user>:<pass>] docker://<registry>/<image>
               Default registries searched: docker.io
               Failed connecting to: docker.io

```

**Comment by clcollins** [comment]
And that got past the check, but fails again, attempting to pull the etcd image from registry.fedoraproject.org:

```
TASK [etcd : Pull new etcd image] *********************************************************************************************************
Thursday 19 April 2018  14:31:43 -0400 (0:00:00.302)       0:01:33.809 ******** 
fatal: [<REDACTED>]: FAILED! => {"changed": true, "cmd": ["docker", "pull", "registry.fedoraproject.org/f26/etcd:3.2.7"], "delta": "0:00:01.158527", "end": "2018-04-19 18:31:45.609830", "msg": "non-zero return code", "rc": 1, "start": "2018-04-19 18:31:44.451303", "stderr": "Error: image f26/etcd:3.2.7 not found", "stderr_lines": ["Error: image f26/etcd:3.2.7 not found"], "stdout": "Trying to pull repository registry.fedoraproject.org/f26/etcd ... \nPulling repository registry.fedoraproject.org/f26/etcd", "stdout_lines": ["Trying to pull repository registry.fedoraproject.org/f26/etcd ... ", "Pulling repository registry.fedoraproject.org/f26/etcd"]}
```

**Comment by michaelgugino** [comment]
Yes, this is currently broken for origin, I haven't had time to fix it yet, PR's welcome.

Related BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1554987

Current workaround: You should be able to pass 'r_etcd_upgrade_version=latest' as an extra_vars when calling ansible-playbook:

```yaml
ansible-playbook -i myinv.txt ... -e "r_etcd_upgrade_version=latest" ...
```

**Comment by clcollins** [comment]
Ah, thanks @michaelgugino.  Before I saw your note, I did some further digging as well, and found the following, too, so it is probably also related to a bad setup on our end.

The call to the fedora etcd image apparently was coming from a play that used a grep and a regex to parse the /etc/systemd/system/etcd_container.service file, in _openshift-ansible/roles/etcd/tasks/upgrade/upgrade_image.yml_:

```
- name: Get current image
  shell: "grep 'ExecStart=' {{ etcd_service_file }} | awk '{print $NF}'"
  register: current_image

- name: Set new_etcd_image
  set_fact:
    new_etcd_image: "{{ current_image.stdout | regex_replace('/etcd.*$','/etcd:' ~ r_etcd_upgrade_version ) }}"

- name: Pull new etcd image
  command: "docker pull {{ new_etcd_image }}"
```

That file apparently had the Fedora etcd image listed in it, for some reason.  I replaced it with the rhel7/etcd image, at that appears to have gotten further.  

So, I guess this was a failure on our end, somehow having a fedora image on the master host.   I'll close this issue.
